initial draft for 2026 project goals (#3935)
rust-lang/rfcs: text/3935-Project-Goals-2026.md

For users, goals serve as a roadmap, giving you an early picture of what work we expect to get done this year.

For Rust maintainers, goals help to surface interactions across teams. They aid in coordination because people know what work others are aiming to do and where they may need to offer support.

## Goals are *proposed* by contributors and *accepted* by teams

As an open-source project, Rust's goal process works differently than a company's. In a company, leadership sets goals and assigns employees to execute them. Rust doesn't have employees - we have contributors who volunteer their time and energy. So in our process, goals begin with the *contributor*: the person (or company) that wants to do the work.

Contributors *propose* goals; Rust teams *accept* them. When you propose a goal, you're saying you're prepared to invest the time to make it happen. When a team accepts, they're committing to support that work - doing reviews, engaging in RFC discussions, and providing the guidance needed to land it in Rust.

## How these goals were selected

> Goal proposals were accepted during the month of January. Many of the goals are continuing goals that are carried over from the previous year, but others goal are new.

Reviewer: ```suggestion
Goal proposals were collected during the month of January. Many of the goals are continuing goals that are carried over from the previous year, but others goal are new.
```
Accepting a proposal sounds ambiguos, did we accept the fact that as a proposal or did we accept the proposal as something we should do.

            map.insert(key, V::default()); //      │
            //  ^~~~~~ ERROR               //      │
            map.get_mut(&key).unwrap()     //      │
        }                                  //      │
    }                                      //      │ ...but today it covers
}                                          // ◄────┘ all this
```

Under Polonius Alpha, this code compiles.

Polonius Alpha is part of a larger roadmap called [the Borrow-Checker Within](https://rust-lang.github.io/rust-project-goals/2026/roadmap-borrow-checker-within.html) that we expect to be driving over the next few years. This year, another part of that work is including Polonius Alpha in [a-mir-formality](https://github.com/rust-lang/a-mir-formality/), the [types team's](https://rust-lang.org/governance/teams/compiler/#team-types) (in-progress) specification for how the Rust type system works. As part of another goal, we are planning to [integrate a-mir-formality into the Rust reference](https://rust-lang.github.io/rust-project-goals/2026/a-mir-formality.html). This would make Polonius the first version of the borrow checker whose behavior is specified outside of the Rust compiler.

> ## Extending const evaluation to *structs/enums*, *traits*, and *reflection*

Reviewer: ```suggestion
## Change const evaluation to support *traits*, and *reflection*, allow *structs/enums* as const parameter types
```
I have no idea what "Extending const evaluation to *structs/enums*" should mean

```rust
fn read_list(path: PathBuf) -> TracedResult<Vec<i32>> {
    let file = File::open(path)?;  // captures location
    Ok(read_number_list(file)?)    // captures location
}
```

No more choosing between readable error handling and useful diagnostics.

The [never type `!`](https://rust-lang.github.io/rust-project-goals/2026/stabilize-never-type.html) has been unstable for *ten years*. It represents computations that never produce a value — like functions that always panic or loop forever. The final blockers are being resolved, and stabilization is in sight.

Finally, the [Sized trait hierarchy](https://rust-lang.github.io/rust-project-goals/2026/scalable-vectors.html) work will stabilize a richer set of sizing traits, which unblocks [extern types](https://github.com/rust-lang/rfcs/pull/1861) — another long-requested feature. Today, `?Sized` conflates "unsized but has metadata" with "truly sizeless." The new hierarchy distinguishes these cases. This same work is also laying the foundation for scalable vector support (Arm SVE), where vector sizes depend on the CPU rather than being fixed at compile time.

> ## Going "beyond the `&`" with better integration for custom pointer types

Reviewer 2: This is missing several of the goals listed under the roadmap: https://rust-lang.github.io/rust-project-goals/2026/roadmap-beyond-the-ampersand.html#how-we-get-there

- Reborrow traits
- In-place init
- Supertrait auto impl

Of the above I'd say the Reborrow traits are 100% this, while the other two are more tangentially related.

| [Arbitrary self types](https://rust-lang.github.io/rust-project-goals/2026/arbitrary-self-types.html#arbitrary-self-types) | Let user-defined smart pointers work as method receivers, stabilizing the `arbitrary_self_types` feature. |
| [`Deref`/`Receiver` split chain experiment](https://rust-lang.github.io/rust-project-goals/2026/arbitrary-self-types.html#deref-receiver-split-chain-experiment) | Experiment with letting `Receiver::Target` and `Deref::Target` diverge, collecting data on utility and use cases. |
| [`derive(CoercePointee)`](https://rust-lang.github.io/rust-project-goals/2026/arbitrary-self-types.html#derive-coercepointee) | Support `dyn Trait` coercion for user-defined smart pointers. |
| [Map the field projection design space](https://rust-lang.github.io/rust-project-goals/2026/field-projections.html#map-the-field-projection-design-space) | Explore the virtual places approach, document it in the beyond-refs wiki, formalize borrow checker integration, and build a compiler experiment. |

*People involved: [Benno Lossin][], [Alice Ryhl][], [Ding Xiang Fei][], [Jack Huey][], [Rémy Rakic][], [Niko Matsakis][], [Oliver Scherer][], [Tyler Mandry][], [TC][]*

---

> Two goals this year are working to make it possible for user-defined types to be used in all the ways that you can use `Box`, `Arc`, and `&`.

Reviewer 2: ```suggestion
Three goals this year are working to make it possible for user-defined types to be used in all the ways that you can use `Box`, `Arc`, and `&`.
```

[Arbitrary self types](https://rust-lang.github.io/rust-project-goals/2026/arbitrary-self-types.html) lets you use custom smart pointers as method receivers. With the `Receiver` trait and `derive(CoercePointee)`, your pointer types work just like `Box` or `Arc` — including method dispatch and coercion to `dyn Trait`:

```rust
impl Person {
    fn biometrics(self: &SmartPointer<Self>) -> &Biometrics {
        ...
    }
}

let person: SmartPointer<Person> = get_data();
let bio = person.biometrics(); // just works
```

> 

Reviewer 2: ```suggestion

[Reborrow traits](https://rust-lang.github.io/rust-project-goals/2026/reborrow-traits.html) allow custom pointers to be reborrowed, just like mutable references. When working with Pin, for example, you should no longer have to call `pinned_ref.as_mut()` to fix lifetime issues.

```

*People involved: [lcnr][], [Niko Matsakis][]*

---

This year, the Rust types team plans to stabilize the [next-generation trait solver](https://rust-lang.github.io/rust-project-goals/2026/next-solver.html). This solver is a ground-up rewrite of the core engine that decides whether types satisfy trait bounds, normalizes associated types, and more. The types team has been working on it since late 2022, and it already powers coherence checking as of Rust 1.84. The goal for this year is to stabilize it for use across all of Rust and remove the old implementation.

This goal may not *sound* like it's going to impact your life, but finishing the new solver unblocks a *lot* of stuff. To start, it allows us to make progress on the [Project Zero](https://rust-lang.github.io/rust-project-goals/2026/roadmap-project-zero.html) roadmap, which aims to fix every known type system soundness bug. It also unblocks long-desired features like implied bounds, cyclic trait matching, and features needed by the [Just add async](https://rust-lang.github.io/rust-project-goals/2026/roadmap-just-add-async.html) roadmap.

> Roadmaps offer a "zoomed out" view of the Rust project direction. Each roadmap collects a set of related project goals into a coherent theme. A typical roadmap takes several years to drive to completion, so when you look at the roadmap, you'll see not only the work we expect to do this year, but a preview of the work we expect in future years (to the extent we know that).

Reviewer: This line feels odd as part of the "## Closing soundness bugs and supporting new lang features with a new trait solver" header

maybe start with either another `---` or add a `Roadmaps` header

Reviewer 3: This looks misplaced, should probably be part of the introduction.

| [Stabilize the next-generation trait solver](https://rust-lang.github.io/rust-project-goals/2026/next-solver.html)                                                                                    | Large  | [lcnr][]     | \*2                    |
| [Arbitrary Self Types](https://rust-lang.github.io/rust-project-goals/2026/arbitrary-self-types.html)                                                                                                 | Large  | [Jack Huey][] | \*3                    |
| [Full Const Generics](https://rust-lang.github.io/rust-project-goals/2026/const-generics.html)                                                                                                        | Large  | [Boxy][]  | \*4                    |
| [Stabilize and model Polonius Alpha](https://rust-lang.github.io/rust-project-goals/2026/polonius.html)                                                                                               | Large  | [Jack Huey][] | \*5                    |

> | [Stabilize concrete type specialization](https://rust-lang.github.io/rust-project-goals/2026/specialization.html)                                                                                     | Large  | [Jack Huey][] | \*6                    |

Reviewer: @rustbot concern specialization-goal

The types team does not accept this goal as is. We've talked about adjusting it, but until then, I consider this blocking.

Reviewer: resolved by  https://github.com/rust-lang/rust-project-goals/pull/612

Reviewer: @rustbot resolve specialization-goal

| [Stabilize Unsafe Fields](https://rust-lang.github.io/rust-project-goals/2026/unsafe-fields.html)                                                        | Small  |                 | \*3 |
| [Continue resolving `cargo-semver-checks` blockers for merging into cargo](https://rust-lang.github.io/rust-project-goals/2026/cargo-semver-checks.html) | Small  | [Alona Enraght-Moony][]  | \*4 |

\*1: Figure out how such API changes should be presented in the API docs. ([from here](https://rust-lang.github.io/rust-project-goals/2026/library-api-evolution.html))

\*2: Design decision and PR review ([from here](https://rust-lang.github.io/rust-project-goals/2026/cargo-script.html))

\*3: Will need approval for rustdoc support. ([from here](https://rust-lang.github.io/rust-project-goals/2026/unsafe-fields.html))

\*4: Discussion and moral support ([from here](https://rust-lang.github.io/rust-project-goals/2026/cargo-semver-checks.html))

#### rustfmt team
| Goal                                        | Level | Champion | Notes |
| :--                                         | :--   | :-- | :-- |

> | [Stabilize Unsafe Fields](https://rust-lang.github.io/rust-project-goals/2026/unsafe-fields.html) | Small |  | \*1 |

Reviewer 4: @ytmimi this is the one ask on the rustfmt team.

It's about approving rustfmt support for allowing the `unsafe` keyword in a struct field definition.

For context, here's the:
project goal: https://rust-lang.github.io/rust-project-goals/2026/unsafe-fields.html
tracking issue: https://github.com/rust-lang/rust/issues/132922
RFC: https://rust-lang.github.io/rfcs/3458-unsafe-fields.html

*[View changes since the review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rfcs/3935/458de400c280de1b839eb3b00ebe2e5151b7c907..ea519a8e23e4b9ae5b54acc3261689cdfbac33b1)*

Reviewer 5: Thank you. I'll take a look at this after work today. After reviewing, how do I give my approval?

Reviewer 3: @ytmimi I think checking your box [here](https://github.com/rust-lang/rfcs/pull/3935#issuecomment-4127175801) should be enough.

Reviewer 3: @ytmimi have you had the chance to look at the ask?

Reviewer 5: @nxsaken My box is now checked ✅

An unstable `-Zbuild-std` flag has existed for a while, but this new design — progressing through a series of RFCs ([one accepted](https://github.com/rust-lang/rfcs/pull/3873), [two more](https://github.com/rust-lang/rfcs/pull/3874) [in review](https://github.com/rust-lang/rfcs/pull/3875)) — has a path to stabilization. Build-std is also part of the [Rust for Linux](https://rust-lang.github.io/rust-project-goals/2026/roadmap-rust-for-linux.html) roadmap.

## Closing soundness bugs and supporting new lang features with a new trait solver

| Goal | What and why |
| --- | --- |
| [Stabilize the next-generation trait solver](https://rust-lang.github.io/rust-project-goals/2026/next-solver.html) | Replace the existing trait solver with a sound, maintainable implementation that unblocks soundness fixes and async features |

*People involved: [lcnr][], [Niko Matsakis][]*

---

> 
> This year, the Rust types team plans to stabilize the [next-generation trait solver](https://rust-lang.github.io/rust-project-goals/2026/next-solver.html). This solver is a ground-up rewrite of the core engine that decides whether types satisfy trait bounds, normalizes associated types, and more. The types team has been working on it since late 2022, and it already powers coherence checking as of Rust 1.84. The goal for this year is to stabilize it for use across all of Rust and remove the old implementation.
> 
> This goal may not *sound* like it's going to impact your life, but finishing the new solver unblocks a *lot* of stuff. To start, it allows us to make progress on the [Project Zero](https://rust-lang.github.io/rust-project-goals/2026/roadmap-project-zero.html) roadmap, which aims to fix every known type system soundness bug. It also unblocks long-desired features like implied bounds, cyclic trait matching, and features needed by the [Just add async](https://rust-lang.github.io/rust-project-goals/2026/roadmap-just-add-async.html) roadmap.
> Roadmaps offer a "zoomed out" view of the Rust project direction. Each roadmap collects a set of related project goals into a coherent theme. A typical roadmap takes several years to drive to completion, so when you look at the roadmap, you'll see not only the work we expect to do this year, but a preview of the work we expect in future years (to the extent we know that).
> 
> ## Active roadmaps
> 
> Not every goal is part of a roadmap, nor are they all expected to be. This initial set of roadmaps is based on the trends that we saw in the 2026 goals. Over time, we expect to add more roadmaps and refine their structure to help people quickly see where Rust is going.

Reviewer 3: ```suggestion
This goal may not *sound* like it's going to impact your life, but finishing the new solver unblocks a *lot* of stuff. To start, it allows us to make progress on the [Project Zero](https://rust-lang.github.io/rust-project-goals/2026/roadmap-project-zero.html) roadmap, which aims to fix every known type system soundness bug. It also unblocks long-desired features like implied bounds, cyclic trait matching, and features needed by the [Just add async](https://rust-lang.github.io/rust-project-goals/2026/roadmap-just-add-async.html) roadmap.

## Active roadmaps

Roadmaps offer a "zoomed out" view of the Rust project direction. Each roadmap collects a set of related project goals into a coherent theme. A typical roadmap takes several years to drive to completion, so when you look at the roadmap, you'll see not only the work we expect to do this year, but a preview of the work we expect in future years (to the extent we know that).

Not every goal is part of a roadmap, nor are they all expected to be. This initial set of roadmaps is based on the trends that we saw in the 2026 goals. Over time, we expect to add more roadmaps and refine their structure to help people quickly see where Rust is going.
```

## Going "beyond the `&`" with better integration for custom pointer types

| Goal | What and why |
| --- | --- |
| [Arbitrary self types](https://rust-lang.github.io/rust-project-goals/2026/arbitrary-self-types.html#arbitrary-self-types) | Let user-defined smart pointers work as method receivers, stabilizing the `arbitrary_self_types` feature. |
| [`Deref`/`Receiver` split chain experiment](https://rust-lang.github.io/rust-project-goals/2026/arbitrary-self-types.html#deref-receiver-split-chain-experiment) | Experiment with letting `Receiver::Target` and `Deref::Target` diverge, collecting data on utility and use cases. |
| [`derive(CoercePointee)`](https://rust-lang.github.io/rust-project-goals/2026/arbitrary-self-types.html#derive-coercepointee) | Support `dyn Trait` coercion for user-defined smart pointers. |
| [Map the field projection design space](https://rust-lang.github.io/rust-project-goals/2026/field-projections.html#map-the-field-projection-design-space) | Explore the virtual places approach, document it in the beyond-refs wiki, formalize borrow checker integration, and build a compiler experiment. |

> *People involved: [Benno Lossin][], [Alice Ryhl][], [Ding Xiang Fei][], [Jack Huey][], [Rémy Rakic][], [Niko Matsakis][], [Oliver Scherer][], [Tyler Mandry][], [TC][]*

Reviewer 3: ```suggestion
*People involved: [Benno Lossin][], [Alice Ryhl][], [Ding Xiang Fei][], [Jack Huey][], [Niko Matsakis][], [Oliver Scherer][], [Tyler Mandry][], [TC][]*
```

|                                                                                            |                   | **[types][]**             | [Oliver Scherer][]          |
|                                                                                            |                   | [compiler][]              | *n/a*             |
| [Architectural groundwork for expansion-time evaluation](https://rust-lang.github.io/rust-project-goals/2026/expansion-time-evaluation.html)     | [Tyler Mandry][]          | **[compiler][]**          | [Vadim Petrochenkov][]     |
|                                                                                            |                   | [types][]                 | [Oliver Scherer][]          |
| [Field Projections](https://rust-lang.github.io/rust-project-goals/2026/field-projections.html)                                                  | [Benno Lossin][]      | **[lang][]**              | [Tyler Mandry][]          |
|                                                                                            |                   | [compiler][]              | [Ding Xiang Fei][] |

> |                                                                                            |                   | [types][]                 | [Rémy Rakic][]              |

Reviewer 3: ```suggestion
|                                                                                            |                   | [types][]                 | [Niko Matsakis][]              |
```

| [lcnr][]             | 5 | [Dictionary Passing Style Experiment](https://rust-lang.github.io/rust-project-goals/2026/dictionary-passing-style-experiment.html)                                                                     |
|                   |   | [Immobile types and guaranteed destructors](https://rust-lang.github.io/rust-project-goals/2026/move-trait.html)                                                                                        |
|                   |   | [Prepare TAIT + RTN for stabilization](https://rust-lang.github.io/rust-project-goals/2026/rtn.html)                                                                                                    |
|                   |   | [Sized Hierarchy and Scalable Vectors](https://rust-lang.github.io/rust-project-goals/2026/scalable-vectors.html)                                                                                       |
|                   |   | [Stabilize the next-generation trait solver](https://rust-lang.github.io/rust-project-goals/2026/next-solver.html)                                                                                      |

> | [Rémy Rakic][]              | 1 | [Field Projections](https://rust-lang.github.io/rust-project-goals/2026/field-projections.html)                                                                                                         |

Reviewer 3: ```suggestion
```

| [Niko Matsakis][]     | 6 | [Box notation for dyn async trait](https://rust-lang.github.io/rust-project-goals/2026/afidt-box.html)                                                                                                  |
|                   |   | [Ergonomic ref-counting](https://rust-lang.github.io/rust-project-goals/2026/ergonomic-rc.html)                                                                                                         |
|                   |   | [Full Const Generics](https://rust-lang.github.io/rust-project-goals/2026/const-generics.html)                                                                                                          |
|                   |   | [Sized Hierarchy and Scalable Vectors](https://rust-lang.github.io/rust-project-goals/2026/scalable-vectors.html)                                                                                       |
|                   |   | [Stabilize Unsafe Fields](https://rust-lang.github.io/rust-project-goals/2026/unsafe-fields.html)                                                                                                       |

> |                   |   | [Stabilize the next-generation trait solver](https://rust-lang.github.io/rust-project-goals/2026/next-solver.html)                                                                                      |

Reviewer 3: ```suggestion
| [Niko Matsakis][]     | 7 | [Box notation for dyn async trait](https://rust-lang.github.io/rust-project-goals/2026/afidt-box.html)                                                                                                  |
|                   |   | [Ergonomic ref-counting](https://rust-lang.github.io/rust-project-goals/2026/ergonomic-rc.html)                                                                                                         |
|                   |   | [Field Projections](https://rust-lang.github.io/rust-project-goals/2026/field-projections.html)                                                                                                          |
|                   |   | [Full Const Generics](https://rust-lang.github.io/rust-project-goals/2026/const-generics.html)                                                                                                          |
|                   |   | [Sized Hierarchy and Scalable Vectors](https://rust-lang.github.io/rust-project-goals/2026/scalable-vectors.html)                                                                                       |
|                   |   | [Stabilize Unsafe Fields](https://rust-lang.github.io/rust-project-goals/2026/unsafe-fields.html)                                                                                                       |
|                   |   | [Stabilize the next-generation trait solver](https://rust-lang.github.io/rust-project-goals/2026/next-solver.html)                                                                                      |
```

| [Stabilize and model Polonius Alpha](https://rust-lang.github.io/rust-project-goals/2026/polonius.html)                                                                                               | Large  | [Jack Huey][] | \*5                    |
| [Stabilize concrete type specialization](https://rust-lang.github.io/rust-project-goals/2026/specialization.html)                                                                                     | Large  | [Jack Huey][] | \*6                    |
| [Immobile types and guaranteed destructors](https://rust-lang.github.io/rust-project-goals/2026/move-trait.html)                                                                                      | Large  | [lcnr][]     | \*7                    |
| [Assumptions on Binders](https://rust-lang.github.io/rust-project-goals/2026/assumptions_on_binders.html)                                                                                             | Medium | [Boxy][]  | \*8                    |

> | [Field Projections](https://rust-lang.github.io/rust-project-goals/2026/field-projections.html)                                                                                                       | Medium | [Rémy Rakic][]      | \*9                    |

Reviewer 3: ```suggestion
| [Field Projections](https://rust-lang.github.io/rust-project-goals/2026/field-projections.html)                                                                                                       | Medium | [Niko Matsakis][]      | \*9                    |
```
